### DISCUSSION TITLE
🐞 RealContentHashPlugin breaks content of assets

### DISCUSSION THREAD

**Discussion by maksimr** [webpack/webpack#14776] [General]
In the STR below, I'll use an intentionally **small** number just to simplify reproduce the problem. But in our project, we stuck with this issue for a reasonable number **4**

STR:

- Clone [the project](https://github.com/maksimr/webpack-14775)
- Run `npm ci`
- Run `npm run build`
- Open `open dist/index.html`

AR: HTML asset corrupted by [replacing hash algorithm](https://github.com/webpack/webpack/blob/c181294865dca01b28e6e316636fef5f2aad4eb6/lib/optimize/RealContentHashPlugin.js#L301) in asset's content 

<img width="900" alt="image" src="https://user-images.githubusercontent.com/616193/142657246-c1516b91-987e-4c84-9eb3-703f7be4aa46.png">
 

**Other relevant information:**
webpack version:  5.64.1
Node.js version: v16.13.0
Operating System: macOS
Additional tools:

**Comment by maksimr** [comment]
@alexander-akait does it work as expected?

**Reply by alexander-akait** [reply]
Sorry for delay, yes, it is bug/limitation, don't use `hashDigestLength: 1`, we already have the issue about it https://github.com/webpack/webpack/issues/14058, it sounds like different, but the roots of them problem are same (here case with `2`). Honestly, it's not very easy to fix it. We have ideas, ideally each file in compilation should provide places for replacement (offset/line/column), but this will require changes from plugins side too. Also `hashDigestLength: 1` is potentially unsafe even we fix it, because you will have collisions on medium/big project.

Feel free to feedback

**Reply by maksimr** [reply]
@alexander-akait thanks a lot for the response.

> don't use hashDigestLength: 1

Sure, as I mentioned in the description, it *only for reproducing problem purpose*. In our project, we had this problem for a hash length **4**, and this length was in *path expression*, not the `hashDigestLength` property itself. Which make this problem even harder to track/debug.

> Honestly, it's not very easy to fix it

I see and understand that figure out right fix will be not so easy. However, we make fix on our side by increasing length, so it's not so *hot* for us and we totally can wait and deal with it.

**Reply by maksimr** [reply]
>We have ideas, ideally each file in compilation should provide places for replacement (offset/line/column)

What about replacing not the *hash* itself but path expression `[name].[hash].[extension]`? Because I don't see when we will use only *hash* in content without asset's path

**Reply by alexander-akait** [reply]
> Sure, as I mentioned in the description, it only for reproducing problem purpose. In our project, we had this problem for a hash length 4, and this length was in path expression, not the hashDigestLength property itself. Which make this problem even harder to track/debug.

Yep, our recommendation to use `8` as minimum 

> What about replacing not the hash itself but path expression [name].[hash].[extension]? Because I don't see when we will use only hash in content without asset's path

It is impossible/hard due circular dependencies, anyway we are looking for solution(s)
